### ISSUE TITLE
Pass all props to a custom snackbar

### ISSUE THREAD

**Issue by MarkusFriesen** [iamhosseindhv/notistack#242]
##  Expected Behavior
I want the options that I pass down calling `enqueueSnackbar` to be passed down to the Custom Snackbar within the `SnackbarProvider`.

Ex: 
calling: 
```jsx
this.props.enqueueSnackbar('An error occured. Try again later.', {header: 'Timeout', variant: 'error'})
``` 
should pass the options down to: 
```jsx
<SnackbarProvider  
    maxSnack={3}
    content={(key, message, options) => <Message id={key} content={message} {...options}/>}
/>
```

## Current Behavior
Currently the options parameter is omitted. 
```jsx
<SnackbarProvider  
    maxSnack={3}
    content={(key, message, ~options~) => <Message id={key} content={message} />}
/>
```

## Steps to Reproduce

See the code above :) 

## Context

I need more context in my custom Snackbar, to change the style and content that is independent of the message

## Your Environment
Independent of environment.

| Tech         | Version |
|--------------|---------|
| Notistack | v0.9.9  |

**Comment by vladk96** [comment]
@MarkusFriesen  agrees with you.
In our project, we need to customize our Snackbar as well because we have different notification styles. So we need to get access to that `options` object.

**Comment by dvakatsiienko** [comment]
Joining the discussion, we need this functionality indeed!

**Comment by orangeswim** [comment]
In the snackbar API, there is "ContentProps" that we should be able to use to pass data into the custom content or SnackbarContent component

**Comment by iamhosseindhv** [comment]
Duplicate of https://github.com/iamhosseindhv/notistack/issues/218

**Comment by SimbaNetropolix** [comment]
To anyone still at it, my hacky solution was to serialize my JSON message into a json string as i created an object of both the message and options and then Parse the string in the custom snackbar component.

**Comment by Bdthomson** [comment]
Can we not just pass in `snack.options` here?
https://github.com/iamhosseindhv/notistack/blob/master/src/SnackbarItem/SnackbarItem.tsx#L167

**Comment by BrentFarese** [comment]
I have been using the `content` key of the [mutual props of the API](https://iamhosseindhv.com/notistack/api#mutual) to provide a complete custom snackbar to Notistack with some success.  I've seen other suggestions, in the interim before this feature is ready, to provide props to a custom snackbar.  For instance, passing a stringified JSON object through the `message` prop of `SnackbarProvider` or other work arounds.

@iamhosseindhv would you consider using the `content` key safe in the way I am doing until #218 is ready?  I've tested it a few ways and it seems to work well so far.  Thank you!

Ex:

```
const key = uuid4();

enqueueSnackbar(fakeNotistackMessage, {
                content: (
                  <CustomSnack id={key} closeSnackbar={closeSnackbar} instanceId={instanceId} />
                ),
                key,
                persist: true
              });
```

**Comment by iamhosseindhv** [comment]
It's been a while.

Download notistack v3 using `npm i notistack@alpha` (in alpha) which gives you full control. Instead of trying to override styles, you make your own snackbars. Here's an example: https://codesandbox.io/s/pedantic-shaw-y3nis
Take a look at `ThemeResponsiveSnackbar.tsx` and change the styles as you wish.

On v3, you can define custom variants:

```tsx
<SnackbarProvider
    Components={{
        success: MyCustomSuccessNotification,
        reportComplete: ReportComplete,
    }}
>
</SnackbarProvider>

interface ReportCompleteProps extends CustomContentProps {
    allowDownload: boolean;
}

const ReportComplete = React.forwardRef((props: ReportCompleteProps, ref) => {
    const {
        // You have access to notistack props, options 👇🏼
        variant,
        message
        // as well as your own custom props 👇🏼
        allowDownload,
    } = props;

    // 
})

enqueueSnackbar('Your report is ready to download', {
   persist: true,
   variant: 'reportComplete',
   allowDownload: true, // <-- pass any additional options
})

// ... or
enqueueSnackbar({
   variant: 'reportComplete',
   message: 'Your report is ready to download',
   allowDownload: true, // <-- pass any additional options
})
```

**Comment by yadielar** [comment]
> Duplicate of #218
> 
> A note for myself to take the points in #218 into account.

@iamhosseindhv I wanted to provide another point for you to take into account when incorporating this request into v3. In addition to passing all the props needed for a custom snack content, it would be very useful for the library to also export the default `SnackbarContent` component so that we can [wrap it with Emotion as I describe in this issue](https://github.com/iamhosseindhv/notistack/issues/461). This would provide a convenient option with the same spirit as the current classes prop, for those who want to customize the default SnackbarContent only slightly without needing to create completely custom components for each variant.

**Comment by amirhossein-shakeri-work** [comment]
I literally need this option:
```js
<SnackbarProvider  
    maxSnack={3}
    content={(key, message, ~options~) => <Message id={key} content={message} />}
/>
```
as its so simpler to use & no need to define new variants compared to this solution
```js
<SnackbarProvider
    Components={{
        success: MyCustomSuccessNotification,
        reportComplete: ReportComplete,
    }}
>
</SnackbarProvider>

interface ReportCompleteProps extends CustomContentProps {
    allowDownload: boolean;
}

const ReportComplete = React.forwardRef((props: ReportCompleteProps, ref) => {
    const {
        // You have access to notistack props, options 👇🏼
        variant,
        message
        // as well as your own custom props 👇🏼
        allowDownload,
    } = props;
})
```

& by the way the main idea of this issue is the first snippet! 

So I would request this option to be added in v3 too!

**Comment by Abolfazl2647** [comment]
hi, i don't know which is better after all.
it would be nice if i somehow i know which type of snackbar im going to show in the props of `MyNotiComponent`.
i just know id and message plus style with visibility of undefined. ???!!!

`
<SnackbarProvider
    content={(id, message) => <MyNotiComponent />}
>
</SnackbarProvider>
`

**Comment by Abolfazl2647** [comment]
hi I created a [PR](https://github.com/iamhosseindhv/notistack/pull/464) for this issue hope it works.

**Comment by iamhosseindhv** [comment]
@amirhossein-shakeri-work I don't think you've done a fair comparison of the two approaches. Having access to additional/custom props only makes sense if one has created a custom component. In your example, you have created a custom component called `Message`.

So really here's the difference between the two approaches:
```jsx
// your proposed approach

<SnackbarProvider  
    content={(key, message, ~options~) => <Message id={key} content={message} />}
/>
```

```jsx
// what's been implemented in v3

<SnackbarProvider
    Components={{
        default: Message,
    }}
>
</SnackbarProvider>
```

Note that you don't have to default a **_new_** variant. Here I'm overriding the default one.

**Comment by iamhosseindhv** [comment]
@yadielar I agree it'd be nice to allow customisation through emotion for those who want to slightly customise the appearance of snackbars.  We'll pick this up in the issue you've opened.

**Comment by nichita-pasecinic** [comment]
@iamhosseindhv the provided [codesandbox example](https://codesandbox.io/s/pedantic-shaw-y3nis) has a bug described in https://github.com/iamhosseindhv/notistack/issues/496

The Snackbars are being closed all at once (after first Snackbar timer expired I presume)
The `persist: true` and `autoHideDuration={xxxxx}` doesn't really help much to solve the issue

**Comment by iamhosseindhv** [comment]
@nichita-pasecinic This has been fixed #495 . Here's an example: https://codesandbox.io/s/twilight-https-xgto47?file=/src/Snackbar.js

Regarding this ticket, head over to [custom-component page](https://notistack.com/examples/advanced/custom-component) of the documentation website to see how you can access all options/props in your custom component.

At the time of writing this, the solution in the documentation website is available on notistack v3  (`npm i notistack@alpha`) which is in alpha stage. I'll be publishing it as stable soon.

**Comment by tetchel** [comment]
this really isn't finished, more like worked around. 

implementing a custom component isn't the same as the original issue request, it's much more complicated.
